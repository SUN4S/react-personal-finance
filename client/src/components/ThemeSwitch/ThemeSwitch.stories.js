import { ThemeSwitch } from "./ThemeSwitch";

export default {
  title: "Theme Switch",
  component: ThemeSwitch,
};

export const Default_Light = () => (
  <html data-theme="light">
    <ThemeSwitch />
  </html>
);

export const Default_Dark = () => (
  <html data-theme="dark">
    <ThemeSwitch />
  </html>
);
