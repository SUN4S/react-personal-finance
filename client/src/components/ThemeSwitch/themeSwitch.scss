#themeButton {
  width: 26px;
  display: flex;
  align-items: center;
  background-color: transparent;
  transition: all 0.3s ease-in-out;

  svg {
    height: 26px;
    width: 26px;
  }
}
