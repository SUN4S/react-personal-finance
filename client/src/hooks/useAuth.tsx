import { useEffect, useState } from "react";

import { useIsLoggedInQuery } from "../services/user";

export const useAuth = () => {
  const { isSuccess } = useIsLoggedInQuery({});

  const getStatus = () => {
    if (isSuccess) {
      return true;
    } else {
      return false;
    }
  };

  return { getStatus };
};
