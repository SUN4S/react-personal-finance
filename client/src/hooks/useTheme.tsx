import { useEffect, useState } from "react";

// Custom hook to set/get data abount current theme
export const useTheme = () => {
  // set current theme bu checking if it exists in local storage
  // if it DOES NOT exist, set it to "light" by default
  const [currentTheme, setCurrentTheme] = useState<string | null>(
    localStorage.getItem("theme") ? localStorage.getItem("theme") : "light"
  );

  // Setting a different theme also store data into local storage
  const setTheme = (mode: string) => {
    localStorage.setItem("theme", mode);
    setCurrentTheme(mode);
  };

  // On load, set html->data-theme to currentTheme('light' or 'dark')
  useEffect(() => {
    currentTheme &&
      document.documentElement.setAttribute("data-theme", currentTheme);
  }, [currentTheme]);

  return { currentTheme, setTheme };
};
