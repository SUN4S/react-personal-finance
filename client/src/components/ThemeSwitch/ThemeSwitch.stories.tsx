import { Meta, Story } from "@storybook/react/types-6-0";

import { ThemeSwitch } from "./ThemeSwitch";

interface TemplateProps {
  theme: string;
  currentTheme: string;
  function: Function;
}

export default {
  title: "Components/Theme Switch",
  component: ThemeSwitch,
} as Meta;

const Template: Story<TemplateProps> = (args) => {
  return (
    <html data-theme={args.theme}>
      <body
        style={{
          width: "400px",
          height: "100vh",
          padding: "8px",
          backgroundColor: "var(--page-bg)",
        }}
      >
        <ThemeSwitch theme={args.currentTheme} clickFunction={args.function} />
      </body>
    </html>
  );
};

export const Default_Light = Template.bind({});
Default_Light.args = {
  theme: "light",
  currentTheme: "light",
  function: () => alert("clicked"),
};

export const Default_Dark = Template.bind({});
Default_Dark.args = {
  theme: "dark",
  currentTheme: "light",
  function: () => alert("clicked"),
};
