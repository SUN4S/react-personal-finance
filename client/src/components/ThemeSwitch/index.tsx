import "./themeSwitch.scss";

import React, { ChangeEvent, useEffect, useState } from "react";

export const ThemeSwitch = () => {
  // Check if theme has at some point been set to localstorage
  // If not, set currentTheme state to 'light'
  const [currentTheme, setCurrentTheme] = useState<string | null>(
    localStorage.getItem("theme") ? localStorage.getItem("theme") : "light"
  );

  // On load, set html->data-theme to currentTheme('light' or 'dark')
  useEffect(() => {
    currentTheme &&
      document.documentElement.setAttribute("data-theme", currentTheme);
  }, [currentTheme]);

  // Function that handles theme change
  // Inverts selected option,
  // sets state to that option
  // and saves state to localStorage for future use
  const handleThemeChange = () => {
    if (currentTheme === "light") {
      document.documentElement.setAttribute("data-theme", "dark");
      setCurrentTheme("dark");
      localStorage.setItem("theme", "dark"); //add this
    } else {
      document.documentElement.setAttribute("data-theme", "light");
      setCurrentTheme("light");
      localStorage.setItem("theme", "light"); //add this
    }
  };

  return (
    <div className="themeContainer">
      <button aria-label="Theme switch" onClick={handleThemeChange}>
        {currentTheme === "dark" ? (
          <i className="fa-solid fa-moon"></i>
        ) : (
          <i className="fa-solid fa-sun"></i>
        )}
      </button>
    </div>
  );
};
