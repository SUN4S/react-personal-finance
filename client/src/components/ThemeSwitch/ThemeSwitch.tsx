import "./ThemeSwitch.scss";

import { IconMoon } from "../Icons/IconMoon/IconMoon";
import { IconSun } from "../Icons/IconSun/IconSun";
import { useTheme } from "../../hooks/useTheme";

export const ThemeSwitch = (props: {
  theme: string | null;
  clickFunction: Function;
}) => {
  return (
    <button
      id="themeButton"
      data-testid="themeButton"
      aria-label="Theme switch"
      onClick={() => props.clickFunction()}
    >
      {props.theme === "dark" ? <IconMoon /> : <IconSun />}
    </button>
  );
};
