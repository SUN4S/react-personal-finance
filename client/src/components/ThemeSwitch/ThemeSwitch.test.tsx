import {
  act,
  cleanup,
  fireEvent,
  render,
  renderHook,
} from "@testing-library/react";

import { ThemeSwitch } from "./ThemeSwitch";
import renderer from "react-test-renderer";
import { useTheme } from "../../hooks/useTheme";

describe("Theme Switch", () => {
  afterEach(cleanup);
  const mockFunction = jest.fn();

  it("Renders", () => {
    const { getByTestId } = render(
      <ThemeSwitch theme="light" clickFunction={mockFunction} />
    );
    expect(getByTestId("themeButton")).toBeTruthy();
  });

  it("Is Clickable", () => {
    const { getByTestId } = render(
      <ThemeSwitch theme="light" clickFunction={mockFunction} />
    );

    fireEvent.click(getByTestId("themeButton"));
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });

  it("Has Aria-label", () => {
    const { getByTestId } = render(
      <ThemeSwitch theme="light" clickFunction={mockFunction} />
    );

    expect(getByTestId("themeButton")).toHaveAttribute(
      "aria-label",
      "Theme switch"
    );
  });

  it("Renders Sun Svg", () => {
    const { getByTestId } = render(
      <ThemeSwitch theme="light" clickFunction={mockFunction} />
    );

    expect(
      getByTestId("themeButton").querySelector("svg#sunLogo")
    ).toBeTruthy();
  });

  it("Renders Moon Svg", () => {
    const { getByTestId } = render(
      <ThemeSwitch theme="dark" clickFunction={mockFunction} />
    );

    expect(
      getByTestId("themeButton").querySelector("svg#moonLogo")
    ).toBeTruthy();
  });

  it("Matches Light Snapshot", () => {
    const tree = renderer.create(
      <ThemeSwitch theme="light" clickFunction={mockFunction} />
    );
    expect(tree).toMatchSnapshot();
  });

  it("Matches Dark Snapshot", () => {
    const tree = renderer.create(
      <ThemeSwitch theme="dark" clickFunction={mockFunction} />
    );
    expect(tree).toMatchSnapshot();
  });
});
