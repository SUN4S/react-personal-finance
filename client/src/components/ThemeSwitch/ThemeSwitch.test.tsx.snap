// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme Switch Matches Dark Snapshot 1`] = `
<button
  aria-label="Theme switch"
  data-testid="themeButton"
  id="themeButton"
  onClick={[Function]}
>
  <svg
    id="moonLogo"
    viewBox="0 0 38.25 39.91"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      data-name="Layer 2"
      id="Layer_2"
    >
      <g
        data-name="Layer 1"
        id="Layer_1-2"
      >
        <path
          className="cls-1"
          d="M13.88,1A19.54,19.54,0,1,0,37.25,29.5"
        />
        <path
          className="cls-1"
          d="M13.88,1A19.54,19.54,0,0,0,30.59,30.67a19.35,19.35,0,0,0,6.66-1.17"
        />
      </g>
    </g>
  </svg>
</button>
`;

exports[`Theme Switch Matches Light Snapshot 1`] = `
<button
  aria-label="Theme switch"
  data-testid="themeButton"
  id="themeButton"
  onClick={[Function]}
>
  <svg
    id="sunLogo"
    viewBox="0 0 43.08 43.08"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      data-name="Layer 2"
      id="Layer_2"
    >
      <g
        data-name="Layer 1"
        id="Layer_1-2"
      >
        <circle
          className="cls-1"
          cx="21.54"
          cy="21.54"
          r="9.24"
        />
        <line
          className="cls-1"
          x1="21.54"
          x2="21.54"
          y1="8.48"
          y2="1"
        />
        <line
          className="cls-1"
          x1="21.54"
          x2="21.54"
          y1="42.08"
          y2="34.61"
        />
        <line
          className="cls-1"
          x1="34.61"
          x2="42.08"
          y1="21.98"
          y2="21.98"
        />
        <line
          className="cls-1"
          x1="1"
          x2="8.48"
          y1="21.98"
          y2="21.98"
        />
        <line
          className="cls-1"
          x1="30.78"
          x2="36.06"
          y1="12.75"
          y2="7.46"
        />
        <line
          className="cls-1"
          x1="7.02"
          x2="12.3"
          y1="36.51"
          y2="31.22"
        />
        <line
          className="cls-1"
          x1="30.78"
          x2="36.06"
          y1="31.22"
          y2="36.51"
        />
        <line
          className="cls-1"
          x1="7.02"
          x2="12.3"
          y1="7.46"
          y2="12.75"
        />
      </g>
    </g>
  </svg>
</button>
`;
